tests/rptest/tests/cluster_config_test.py

```diff
@@ -1739,3 +1739,85 @@ def assert_restart_status(expect: bool):
         status = self.admin.get_cluster_config_status()
         for n in status:
             assert n['restart'] is False
+
+
+class ClusterConfigLegacyDefaultTest(RedpandaTest, ClusterConfigHelpersMixin):
+    """
+    Test config::legacy_default feature, that defaults for features can be
+    dependent on the original version of an upgraded cluster.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.admin = Admin(self.redpanda)
+        self.rpk = RpkTool(self.redpanda)
+        self.installer = self.redpanda._installer
+
+        self.legacy_version = (23, 1)
+
+        # this is synchronized with configuration.cc
+        self.key = 'topic_partitions_per_shard'
+        self.legacy_default = 7000
+        self.new_default = 1000
+
+    def setUp(self):
+        pass
+
+    def _upgrade(self, wipe_cache):
+        self.installer.install(self.redpanda.nodes, RedpandaInstaller.HEAD)
+        for node in self.redpanda.nodes:
+            self.redpanda.stop_node(node)
+
+            if wipe_cache:
+                # Erase the cluster config cache, so that the node will replay from
+                # controller log and/or controller snapshots
+                cache_path = f"{self.redpanda.DATA_DIR}/config_cache.yaml"
+                self.logger.info(
+                    "Erasing config cache on {node.name} at {cache_path}")
+                assert node.account.exists(cache_path)
+                node.account.remove(cache_path)
+
+            self.redpanda.start_node(node)
+
+    @cluster(num_nodes=3)
+    @parametrize(wipe_cache=True)
+    @parametrize(wipe_cache=False)
+    def test_legacy_default(self, wipe_cache: bool):
+        old_version, _ = self.installer.latest_for_line(self.legacy_version)
+        self.installer.install(self.redpanda.nodes, old_version)
+        self.redpanda.start()
+
+        self._check_value_everywhere(self.key, self.legacy_default)
+        self._upgrade(wipe_cache)
+        self._check_value_everywhere(self.key, self.legacy_default)
```

Comment:
nit: might be worthwhile to restart the nodes and do these checks again, though offhand I don't recall whether this would be an interesting case (not sure where along controller replay configs get set, etc. perhaps this is already covered by the upgrade anyway)

Reply:
made a follow up ticket and will do this I think it is a stronger check